sdk/core/azure-core/azure/core/pipeline/policies/_authentication_async.py

```diff
@@ -92,28 +92,28 @@ async def send(
         await await_result(self.on_request, request)
         try:
             response = await self.next.send(request)
-            await await_result(self.on_response, request, response)
         except Exception:  # pylint:disable=broad-except
-            handled = await await_result(self.on_exception, request)
-            if not handled:
-                raise
+            await await_result(self.on_exception, request)
```

Comment:
Pyright started here by "possible unbound variable 'response'".

The root cause is complex:
- This policy used to be a SansIOPolicy, but got converted to regular policy when challenge was added
- To be backward compatible with someone possibly subclassing it, we kept the on_request/on_response/on_exception methods you would have gotten by inheriting SansIOPolicy. This also means that the implementation of `send` now has to take care of calling those.
- This was implemented incorrectly compared to what the AsyncSansIORunner is actually doing. On_exception never returned a boolean, and so checking if it was handled or not is not a thing. Also, on_response and on_exception are mutually exclusive in the reference implementation, while the code as it was written could have executed both.

As this was added for backward compat and nobody realized we were actually not backward compatible, I would tend to assume that no-one tried to subclass this and relied on SansIO contract.

Reply:
https://github.com/Azure/azure-sdk-for-python/issues/31836

Reply:
When you say `on_response and on_exception are mutually exclusive in the reference implementation` - you mean in terms of documentation?
I'm just wondering if anyone could have been (rightly or wrongly) using the exception handler as a fallback to errors happening in their response handler...

Reply:
I meant this:
https://github.com/Azure/azure-sdk-for-python/blob/df7a56d1d15d2277561d41eb5c5a08b87410b61e/sdk/core/azure-core/azure/core/pipeline/_base_async.py#L68-L75

As you can see the `on_response` is not included in the try catch, and you can never execute both the `on_response` and the `on_exception` at once in the SansIO Runner